miden-lib/asm/sat/epilogue.masm

```diff
@@ -0,0 +1,229 @@
+use.std::sys
+
+use.miden::sat::utils
+
+# CONSTANTS
+# =================================================================================================
+
+# BOOKKEEPING
+# -------------------------------------------------------------------------------------------------
+
+# The memory address at which the number of created notes is stored.
+const.NUM_CREATED_NOTES_PTR=2
+
+# GLOBAL MEMORY
+# -------------------------------------------------------------------------------------------------
+
+# The memory address at which the initial nonce is stored.
+const.INIT_NONCE=14
+
+# CREATED NOTES
+# -------------------------------------------------------------------------------------------------
+
+# The memory address at which the created notes section begins.
+const.CREATED_NOTE_SECTION_OFFSET=10000
+
+# The offsets at which data of a created note is stored relative to the start of its data segment.
+const.CREATED_NOTE_META_OFFSET=0
+const.CREATED_NOTE_RECIPIENT_OFFSET=1
+const.CREATED_NOTE_VAULT_HASH_OFFSET=2
+const.CREATED_NOTE_NUM_ASSETS_OFFSET=3
+const.CREATED_NOTE_ASSETS_OFFSET=4
+
+# CREATED NOTES PROCEDURES
+# =================================================================================================
+
+# TODO: Consider changing the crated notes procedures to work with pointers. Instead of taking an
+# index as input, we could take a pointer to the created note data. This would allow us to avoid
+# the need to compute the pointer multiple times across the different procedures.
+
+#! Computes the vault hash of the created note with index i. The hash is computed as a sequential
+#! hash of the assets contained in the note. If there are an odd number of assets, then for the
+#! final hashing permutation we pad the last word of the hasher rate with ZERO's.
+#!
+#! Stack: [i]
+#! Output: [VAULT_HASH]
+#!
+#! - i is the index of the created note.
+#! - VAULT_HASH is the vault hash of the created note with index i.
+proc.compute_created_note_vault_hash
+    # get created note data pointer from created note index
+    exec.utils::created_note_data_ptr
+
+    # duplicate created note pointer - we will use this to save vault hash to memory
+    dup
+
+    # get the number of assets from memory
+    dup push.CREATED_NOTE_NUM_ASSETS_OFFSET add mem_load
+
+    # round up the number of assets to the next multiple of 2 (simplifies reading of assets)
+    dup is_odd
+
+    # pad to multiple of two if odd
+    if.true
+        push.1 add
+    end
```

Comment:
Instead of doing this, would it make sense to divide by 2 (using integer division) and then just increment by 1 on line  84? This way, the counter keeps track of pairs of assets.

Reply:
Yes I think we could do this as well.  However, I think we will have to increment by 1 first as `u32checked_div` takes the floor value. Some like:
```
pairs = (x + 1) / 2
```

```
pairs:
# [x]
push.1 add #[x + 1]
u32checked_div #[(x+1) / 2]
```

Whats the reason for this proposal? I notice that it eliminates the conditional `if` block which reduces hashing which is good.  I suppose this is more efficient?

Reply:
Yes, should be more efficient. The `if.true` block will actually introduce several nodes into the MAST and here we can avoid this with an instruction which takes just a couple of cycles to execute. Also, incrementing by one can be done in  1 cycles (though, this is a minor point).

Reply:
thanks I've made the update and also updates instance where we increment to use `add.1`.